plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -622,6 +627,34 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
                 .build();
     }
 
+    @Override
+    public void validateScan(ConnectorSession session, ConnectorTableHandle handle)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) handle;
+        Optional<PartitionSpec> partitionSpec = table.getPartitionSpecJson()
+                .map(partitionSpecJson -> PartitionSpecParser.fromJson(SchemaParser.fromJson(table.getTableSchemaJson()), partitionSpecJson));
+        if (partitionSpec.isEmpty() || partitionSpec.get().isUnpartitioned()) {
+            return;
+        }
+        if (isFilterOnPartitionTable(session)
+                && table.getEnforcedPredicate().equals(TupleDomain.all())) {
+            Set<IcebergColumnHandle> icebergConstraintColumns = table.getConstraintColumns().stream()
+                    .map(columnHandle -> (IcebergColumnHandle) columnHandle)
+                    .collect(Collectors.toSet());
+            Set<String> par1 = icebergConstraintColumns.stream().map(IcebergColumnHandle::getName).collect(toImmutableSet());
+            Set<String> par2 = partitionSpec.get().fields().stream()
+                    .filter(field -> !(field.transform().isVoid()))
+                    .map(PartitionField::name)
+                    .collect(toImmutableSet());
+            if (Collections.disjoint(par1, par2)) {
```

Comment:
Is Collections.disjoint() enough here? This would return true even if just one partition column has a predicate, but the other partition columns don't. Don't we want a predicate on all partition columns? `columnsWithPredicates.containsAll(partitionColumns)` would make more sense to me. WDYT?

Reply:
I just checked the HiveConfig equivalent and that says
```
 @ConfigDescription("Require filter on at least one partition column")
```
so if we want to remain consistent we can keep it like this

Reply:
But to be honest, if you don't enforce this on the top-level partition column, and your predicate is only on the second, third or any other lower-level partition column, this config won't be that useful.

Reply:
> I just checked the HiveConfig equivalent and that says
> 
> ```
>  @ConfigDescription("Require filter on at least one partition column")
> ```
> 
> so if we want to remain consistent we can keep it like this

I may not want to make this condition too strict, so the query only needs to have a partition field. In the end, it depends on how we define the semantics of this **prohibition**. Here I simply refer to the implementation of hive.

Reply:
> Is Collections.disjoint() enough here? This would return true even if just one partition column has a predicate, but the other partition columns don't. Don't we want a predicate on all partition columns? `columnsWithPredicates.containsAll(partitionColumns)` would make more sense to me. WDYT?

No, if just one partition column has a predicate, but the other partition columns don't. The code doesn't go here, because once a filter condition is enforced (obviously the partition field). We don't think this is a bad query.

```
 if (isQueryPartitionFilterRequired(session)
                && table.getEnforcedPredicate().equals(TupleDomain.all())) {
// ......
}
```

Reply:
> But to be honest, if you don't enforce this on the top-level partition column, and your predicate is only on the second, third or any other lower-level partition column, this config won't be that useful.

Yes. That's right. I will use code `columnsWithPredicates.containsAll(partitionColumns)` and I think it can solve this issue.

Reply:
@marton-bod I thought about it again based on the current actual situation of our company. If the query is required to carry all the partition fields, it seems too strict, because we do have a lot of queries that only use part of the partition fields. What do you think ?

Reply:
I also think requiring all partition columns in the filter is too strict. Suppose I have a table partitioned by `date` and `hour`, and I want to run queries on a specific day, now I need to specify `date = .... and hour >=0 and hour<=23` which seems not very friendly to me.
An alternative is perhaps requiring a prefix of partition columns. But I prefer the origin solution to just require any partition column. IMHO which partition columns must be filtered should be decided from a business perspective and we're not in a good position to do that for users.

Reply:
That's a good point and I'm on the same page with you.